src/librustc_mir/shim.rs

```diff
@@ -259,6 +271,356 @@ impl<'a, 'tcx> DropElaborator<'a, 'tcx> for DropShimElaborator<'a, 'tcx> {
     }
 }
 
+/// Build a `Clone::clone` shim for `self_ty`. Here, `def_id` is `Clone::clone`.
```

Comment:
So this function is pretty big (~350 loc), but it's difficult to split it without writing some code twice or using functions with dozens of args.
Actually the logic for handling arrays (one loop for filling, another one in case of unwinding) takes 200 loc on its own.